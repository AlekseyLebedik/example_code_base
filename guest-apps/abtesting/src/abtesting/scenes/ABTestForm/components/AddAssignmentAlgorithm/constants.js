export const HASH_FUNCTIONS = ['sha256', 'sha512'];
