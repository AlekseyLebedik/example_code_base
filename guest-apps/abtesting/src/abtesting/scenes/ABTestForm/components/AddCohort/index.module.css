.cohortContainer {
  position: relative;
}

.cohortForm {
  margin-top: 15px;
  position: relative;
  padding-bottom: 0px;
}

.cohortProps {
  display: flex;
  align-items: flex-start;
  min-height: 78px;
  margin-top: -10px;
}

.cohortID {
  position: absolute;
  top: 33px;
  right: 85px;
}

.deleteButton {
  padding: 0 !important;
  width: auto !important;
  height: auto !important;
  position: absolute !important;
  right: -9px;
  top: -8px;
  background-color: #fff !important;
  border-radius: 50% !important;
  box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
}

.addButton {
  position: absolute !important;
  right: 0px;
  top: -6px;
}

.cohortInput {
  margin-right: 10px;
  min-width: 72px;
}

.controlCheckbox {
  width: auto !important;
  margin-left: 10px;
}

@media only screen and (max-width: 414px) {
  .cohortInput {
    width: 100% !important;
  }
}

.groupOverrideRow {
  display: table-row-group;
}

.groupOverrideBox {
  display: table-cell;
  vertical-align: middle;
}

.groupOverrideBox1 {
  display: table-cell;
  height: 81px;
}
