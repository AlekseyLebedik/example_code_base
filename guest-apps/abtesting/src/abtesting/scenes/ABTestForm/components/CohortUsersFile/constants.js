import React from 'react';

export const UPLOAD_FILE_HELP = (
  <div style={{ textAlign: 'left' }}>
    Each record in the csv file should contain {'<'}accountType{'>'} and {'<'}
    userID{'>'}. For example:
    <br />
    <i>
      psn,12334567
      <br />
      steam,2317174234108935
    </i>
    <br />
    <br />
    Invalid records will be skipped.
  </div>
);
