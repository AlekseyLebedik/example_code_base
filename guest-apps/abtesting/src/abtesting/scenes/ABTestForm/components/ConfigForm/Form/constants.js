export const FORM_NAME = 'abtesting__config-form';
