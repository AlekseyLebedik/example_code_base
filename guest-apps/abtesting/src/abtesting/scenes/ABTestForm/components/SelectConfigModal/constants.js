export const FORM_NAME = 'abtesting__select-modal';

export const COLUMNS = [
  {
    headerName: 'Name',
    field: 'name',
    checkboxSelection: true,
    filter: 'agTextColumnFilter',
  },
  {
    headerName: 'Created',
    field: 'created',
  },
  {
    headerName: 'Updated',
    field: 'updated',
    sortable: true,
  },
  {
    headerName: 'Service',
    field: 'serviceID',
    filter: true,
  },
];
