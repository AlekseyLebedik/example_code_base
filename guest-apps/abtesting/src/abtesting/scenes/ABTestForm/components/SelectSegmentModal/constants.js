import React from 'react';
import ConfirmActionComponent from 'dw/core/components/ConfirmActionComponent';
import FeatureSwitchesCheck from '@demonware/devzone-core/access/FeatureSwitchesCheck';
import * as fs from '@demonware/devzone-core/access/FeatureSwitchesCheck/featureSwitches';

export const DeleteColumnFormatter = params => {
  const {
    data,
    context: { onRemove, hasDeletePermission },
  } = params;
  return (
    hasDeletePermission && (
      <FeatureSwitchesCheck
        featureSwitches={[fs.ABTESTING_DELETE_SEGMENT]}
        isStaffAllowed={false}
      >
        <ConfirmActionComponent
          component="IconButton"
          tooltip="Remove"
          onClick={() => onRemove([data])}
          confirm={{
            title: 'Confirm Remove',
            confirmMsg: `Are you sure you want to remove segment "${data.name}"?`,
            mainButtonLabel: 'Remove',
            destructive: true,
          }}
        >
          delete
        </ConfirmActionComponent>
      </FeatureSwitchesCheck>
    )
  );
};

function nameComparator(valueA, valueB) {
  if (valueA && valueB) {
    return valueA.toLowerCase().localeCompare(valueB.toLowerCase());
  }
  return null;
}

export const COLUMNS = [
  {
    field: '',
    headerName: '',
    width: 50,
    checkboxSelection: true,
    headerCheckboxSelection: true,
    suppressMenu: true,
    suppressSizeToFit: true,
  },
  {
    headerName: 'Segment ID',
    field: 'segmentID',
    sortable: true,
  },
  {
    headerName: 'Name',
    field: 'name',
    sortable: true,
    comparator: nameComparator,
    filter: 'agTextColumnFilter',
  },
  {
    headerName: '',
    cellRenderer: 'deleteColumnFormatter',
    width: 94,
    suppressMenu: true,
    suppressSizeToFit: true,
  },
];
