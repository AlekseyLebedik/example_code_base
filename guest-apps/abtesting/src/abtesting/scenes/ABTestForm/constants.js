export const FORM_NAME = 'abtesting__create-test-form';
export const FORM_MODE_ENUM = {
  VIEW: 'VIEW',
  NEW: 'NEW',
  UPDATE: 'UPDATE',
  CLONE: 'CLONE',
  PROPAGATE: 'PROPAGATE',
};

export const OVERRIDES_ENVIRONMENTS = ['cert', 'dev'];
