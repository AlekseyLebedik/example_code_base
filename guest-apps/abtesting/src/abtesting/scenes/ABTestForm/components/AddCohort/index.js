import React from 'react';
import PropTypes from 'prop-types';
import get from 'lodash/get';
import { Field, FieldArray, Fields, formValues } from 'redux-form';
import Tooltip from '@material-ui/core/Tooltip';

import Icon from '@material-ui/core/Icon';
import Paper from '@material-ui/core/Paper';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import MenuItem from '@material-ui/core/MenuItem';
import Portal from '@material-ui/core/Portal';
import Typography from '@material-ui/core/Typography';
import Input from 'dw/core/components/FormFields/Input';
import Select from 'dw/core/components/FormFields/Select';
import SelectField from 'dw/core/components/Select';
import * as V from 'dw/core/components/FormFields/validation';
import Checkbox from 'dw/core/components/FormFields/Checkbox';
import ClipboardTooltip from 'dw/core/components/ClipboardTooltip';

import AddAssignmentAlgorithm from '../AddAssignmentAlgorithm';
import CohortUsersFile from '../CohortUsersFile';
import SelectSegment from '../SelectSegment';
import AddTreatment from '../AddTreatment';
import styles from './index.module.css';
import { FORM_MODE_ENUM } from '../../constants';

const minLength3 = V.minLength(3);

const CohortGroupsBase = props => {
  const {
    cohortOverridesContainer,
    cohortIdx,
    input,
    groupsList,
    cohorts,
    disabled,
  } = props;
  const excludeIds = cohorts.reduce(
    (acc, current) => [...acc, ...get(current, 'groups', [])],
    []
  );
  const cohort = cohorts[cohortIdx];
  if (!cohort) return null;
  const currentGroups = cohort && get(cohort, 'groups', []);
  return (
    <Portal container={cohortOverridesContainer}>
      <div className={styles.groupOverrideRow}>
        <div className={styles.groupOverrideBox}>{cohort.name}</div>
        <div className={styles.groupOverrideBox1}>
          <SelectField
            {...input}
            label="override group"
            multiple
            fullWidth
            disabled={disabled}
          >
            {groupsList
              .filter(
                g =>
                  !excludeIds.includes(g.groupID) ||
                  currentGroups.includes(g.groupID)
              )
              .map(group => (
                <MenuItem value={group.groupID} key={group.groupID}>
                  {group.groupName}
                </MenuItem>
              ))}
          </SelectField>
        </div>
      </div>
    </Portal>
  );
};

CohortGroupsBase.propTypes = {
  cohortOverridesContainer: PropTypes.object,
  cohortIdx: PropTypes.number.isRequired,
  input: PropTypes.object.isRequired,
  groupsList: PropTypes.array.isRequired,
  cohorts: PropTypes.array.isRequired,
  disabled: PropTypes.bool.isRequired,
};

CohortGroupsBase.defaultProps = {
  cohortOverridesContainer: undefined,
};

const CohortGroups = formValues('cohorts')(CohortGroupsBase);

const renderSourceFields = fields => {
  const { index, formMode, cohortID } = fields;
  const sourceValue = fields.cohorts[index].source.input.value;
  const disabled =
    formMode === FORM_MODE_ENUM.PROPAGATE || formMode === FORM_MODE_ENUM.VIEW;
  return (
    <>
      <Select
        {...fields.cohorts[index].source}
        label="Source"
        className={styles.cohortInput}
        disabled={disabled}
      >
        <MenuItem key="source-manual" value="manual">
          Manual
        </MenuItem>
        <MenuItem key="source-global" value="global">
          Global
        </MenuItem>
        <MenuItem key="source-segment" value="segment">
          Segment
        </MenuItem>
      </Select>

      {sourceValue && sourceValue === 'manual' && (
        <CohortUsersFile
          cohort={fields.cohorts[index]}
          cohortID={cohortID}
          disabled={disabled}
        />
      )}

      {sourceValue && sourceValue === 'segment' && (
        <SelectSegment
          cohort={fields.cohorts[index]}
          validate={V.required}
          disabled={disabled}
        />
      )}

      {sourceValue && sourceValue !== 'manual' && (
        <>
          <Input
            {...fields.cohorts[index].percent}
            label="% percentage"
            disabled={sourceValue === 'manual' || disabled}
            className={styles.cohortInput}
          />
          <Input
            {...fields.cohorts[index].maxMembers}
            label="Max Players"
            disabled={disabled}
            className={styles.cohortInput}
            hidden
          />
        </>
      )}
      {cohortID && (
        <div className={styles.cohortID}>
          <ClipboardTooltip label="Cohort ID" clipboardText={cohortID}>
            Cohort ID: {cohortID.slice(0, 4)}
            <strong>...</strong>
          </ClipboardTooltip>
        </div>
      )}
    </>
  );
};

const renderCohorts = ({
  fields,
  formMode,
  disabled,
  handleFromTreatmentDates,
  handleToTreatmentDates,
  selectedCatchStart,
  selectedCatchEnd,
  cohortOverridesContainer,
  groupsList,
  displayCohortOverrides,
  hasCohortOverridesPermission,
}) => (
  <>
    {formMode !== FORM_MODE_ENUM.PROPAGATE &&
    formMode !== FORM_MODE_ENUM.VIEW ? (
      <Tooltip title="Add Cohort">
        <IconButton
          className={styles.addButton}
          onClick={() => fields.push({ treatments: [{}] })}
        >
          <Icon>playlist_add</Icon>
        </IconButton>
      </Tooltip>
    ) : null}
    {fields.map((cohort, index) => (
      <Paper className={styles.cohortForm} key={cohort} data-cy={cohort}>
        <AppBar position="static" color="default">
          <Toolbar className={styles.cohortProps}>
            {formMode !== FORM_MODE_ENUM.PROPAGATE &&
            formMode !== FORM_MODE_ENUM.VIEW ? (
              <Tooltip title="Delete Cohort">
                <IconButton
                  color="secondary"
                  className={styles.deleteButton}
                  onClick={() => fields.remove(index)}
                >
                  <Icon className={styles.deleteIcon}>highlight_off</Icon>
                </IconButton>
              </Tooltip>
            ) : null}
            <Field
              className={styles.cohortInput}
              name={`${cohort}.name`}
              component={Input}
              label="Name"
              validate={[V.required, minLength3]}
              disabled={disabled}
            />
            {displayCohortOverrides && (
              <Field
                name={`${cohort}.groups`}
                component={CohortGroups}
                cohortIdx={index}
                cohortOverridesContainer={cohortOverridesContainer}
                groupsList={groupsList}
                disabled={disabled || !hasCohortOverridesPermission}
              />
            )}
            <Fields
              component={renderSourceFields}
              className={styles.cohortInput}
              names={[
                `${cohort}.source`,
                `${cohort}.percent`,
                `${cohort}.maxMembers`,
                `${cohort}.segmentID`,
                `${cohort}.fileData`,
              ]}
              disabled={disabled}
              index={index}
              formMode={formMode}
              cohortID={fields.get(index).cohortID}
            />
          </Toolbar>
        </AppBar>
        <AddTreatment
          formMode={formMode}
          disabled={disabled && formMode !== FORM_MODE_ENUM.PROPAGATE}
          cohort={cohort}
          cohortIndex={index}
          handleToTreatmentDates={handleToTreatmentDates}
          handleFromTreatmentDates={handleFromTreatmentDates}
          selectedCatchStart={selectedCatchStart}
          selectedCatchEnd={selectedCatchEnd}
          cohortName={fields.get(index).name}
        />
        <Field
          className={styles.controlCheckbox}
          selectedClassName={styles.controlCheckboxChecked}
          name={`${cohort}.isControl`}
          component={Checkbox}
          disabled={disabled}
          label="control"
          labelPosition="right"
          color="primary"
        />
      </Paper>
    ))}
  </>
);

renderCohorts.propTypes = {
  handleFromTreatmentDates: PropTypes.func,
  handleToTreatmentDates: PropTypes.func,
  selectedCatchStart: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  selectedCatchEnd: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  fields: PropTypes.object.isRequired,
  formMode: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  groupsList: PropTypes.array.isRequired,
  displayCohortOverrides: PropTypes.bool.isRequired,
  hasCohortOverridesPermission: PropTypes.bool.isRequired,
  cohortOverridesContainer: PropTypes.object,
};

renderCohorts.defaultProps = {
  cohortOverridesContainer: undefined,
  handleFromTreatmentDates: null,
  handleToTreatmentDates: null,
  selectedCatchStart: null,
  selectedCatchEnd: null,
};

const AddCohort = ({
  formMode,
  showDetails,
  disabled,
  handleToTreatmentDates,
  handleFromTreatmentDates,
  selectedCatchStart,
  selectedCatchEnd,
  cohortOverridesContainer,
  groupsList,
  displayCohortOverrides,
  hasCohortOverridesPermission,
}) => (
  <div className={styles.cohortContainer}>
    <Typography variant="h4" hidden={!showDetails}>
      Cohorts &amp; Treatments
    </Typography>
    <AddAssignmentAlgorithm
      visible={showDetails}
      disabled={disabled || formMode === FORM_MODE_ENUM.UPDATE}
    />
    <FieldArray
      name="cohorts"
      props={{
        formMode,
        disabled,
        handleToTreatmentDates,
        handleFromTreatmentDates,
        selectedCatchStart,
        selectedCatchEnd,
        cohortOverridesContainer,
        groupsList,
        displayCohortOverrides,
        hasCohortOverridesPermission,
      }}
      component={renderCohorts}
    />
  </div>
);

AddCohort.propTypes = {
  formMode: PropTypes.string.isRequired,
  handleFromTreatmentDates: PropTypes.func,
  handleToTreatmentDates: PropTypes.func,
  selectedCatchStart: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  selectedCatchEnd: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  showDetails: PropTypes.bool,
  disabled: PropTypes.bool,
  groupsList: PropTypes.array,
  cohortOverridesContainer: PropTypes.object,
  displayCohortOverrides: PropTypes.bool,
  hasCohortOverridesPermission: PropTypes.bool,
};

AddCohort.defaultProps = {
  showDetails: true,
  disabled: false,
  selectedCatchStart: null,
  selectedCatchEnd: null,
  handleFromTreatmentDates: null,
  handleToTreatmentDates: null,
  groupsList: [],
  cohortOverridesContainer: undefined,
  displayCohortOverrides: false,
  hasCohortOverridesPermission: false,
};

export default AddCohort;
