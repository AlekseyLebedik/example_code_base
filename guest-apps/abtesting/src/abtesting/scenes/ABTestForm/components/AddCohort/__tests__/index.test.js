import React from 'react';
import { shallow } from 'enzyme';

import AddCohort from '../index';
import { FORM_MODE_ENUM } from '../../../constants';

describe('ABTesting component AddCohort', () => {
  it('renders component', () => {
    expect(
      shallow(<AddCohort updateTest formMode={FORM_MODE_ENUM.NEW} />)
    ).toMatchSnapshot();
  });
});
