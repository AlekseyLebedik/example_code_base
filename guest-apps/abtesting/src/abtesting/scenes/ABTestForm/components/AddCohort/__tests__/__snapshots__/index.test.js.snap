// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABTesting component AddCohort renders component 1`] = `
<div
  className="cohortContainer"
>
  <WithStyles(ForwardRef(Typography))
    hidden={false}
    variant="h4"
  >
    Cohorts & Treatments
  </WithStyles(ForwardRef(Typography))>
  <AddAssignmentAlgorithm
    disabled={false}
    visible={true}
  />
  <FieldArray
    component={[Function]}
    name="cohorts"
    props={
      Object {
        "cohortOverridesContainer": undefined,
        "disabled": false,
        "displayCohortOverrides": false,
        "formMode": "NEW",
        "groupsList": Array [],
        "handleFromTreatmentDates": null,
        "handleToTreatmentDates": null,
        "hasCohortOverridesPermission": false,
        "selectedCatchEnd": null,
        "selectedCatchStart": null,
      }
    }
  />
</div>
`;
